File: gui/menu/workspace/workspace_extrapolate_variables.cpp
#include "workspace_extrapolate_variables.h"


QIodeMenuWorkspaceExtrapolateVariables::QIodeMenuWorkspaceExtrapolateVariables(const QString& settings_filepath, QWidget* parent, Qt::WindowFlags f) : 
    QIodeSettings(settings_filepath, parent, f)
{
    setupUi(this);

    completer = new QIodeCompleter(false, false, I_VARIABLES, textEdit_variables_list);
    textEdit_variables_list->setCompleter(completer);

    QList<QString> methodsList; 
    for(const std::string& method : v_simulation_initialization) methodsList << QString::fromStdString(method);

    wFrom = new WrapperSampleEdit(label_sample_from->text(), *sampleEdit_sample_from, REQUIRED_FIELD);
    wTo = new WrapperSampleEdit(label_sample_to->text(), *sampleEdit_sample_to, REQUIRED_FIELD);
    wMethod = new WrapperComboBox(label_method->text(), *comboBox_method, REQUIRED_FIELD, methodsList);
    wVarsList = new WrapperQPlainTextEdit(label_variables_list->text(), *textEdit_variables_list, OPTIONAL_FIELD);

    mapFields["From"] = wFrom;
    mapFields["To"] = wTo;
    mapFields["Method"] = wMethod;
    mapFields["VarsList"] = wVarsList;

    // TODO: if possible, find a way to initialize className inside MixingSettings
    // NOTE FOR DEVELOPPERS: we cannot simply call the line below from the constructor of MixingSettings 
    //                       since in that case this refers to MixingSettings and NOT the derived class
    className = QString::fromStdString(typeid(this).name());
    loadSettings();
}

QIodeMenuWorkspaceExtrapolateVariables::~QIodeMenuWorkspaceExtrapolateVariables()
{
    delete wFrom;
    delete wTo;
    delete wMethod;
    delete wVarsList;

    delete completer;
}

void QIodeMenuWorkspaceExtrapolateVariables::extrapolate_variables()
{
    try
    {
        std::string s_from = wFrom->extractAndVerify().toStdString();
        std::string s_to = wTo->extractAndVerify().toStdString();
        EnumSimulationInitialization method = (EnumSimulationInitialization) wMethod->extractAndVerify();
        std::string vars_list = wVarsList->extractAndVerify().toStdString();

        KDBVariables kdb;
        kdb.extrapolate(method, s_from, s_to, vars_list);
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }
    
    this->accept();
}

void QIodeMenuWorkspaceExtrapolateVariables::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/menu/workspace/workspace_extrapolate_variables.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from settings import MixinSettingsDialog
from .ui_workspace_extrapolate_variables import Ui_MenuWorkspaceExtrapolateVariables


from iode import IodeType, SimulationInitialization, variables


class MenuWorkspaceExtrapolateVariables(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuWorkspaceExtrapolateVariables()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.ui.textEdit_variables_list.handle_iode_type(IodeType.VARIABLES)        
        self.ui.textEdit_variables_list.include_iode_command(False)        
        self.ui.textEdit_variables_list.include_lec_functions(False)        

        self.v_simulation_initialization = list(SimulationInitialization)
        v_simulation_initialization_names = [v.name.upper() for v in self.v_simulation_initialization]
        self.ui.comboBox_method.addItems(v_simulation_initialization_names)

    @Slot()
    def extrapolate_variables(self):
        try:
            from_period: str = self.ui.sampleEdit_sample_from.text() 
            to_period: str = self.ui.sampleEdit_sample_to.text() 
            i_simu_method: int = self.ui.comboBox_method.currentIndex()
            simu_method: SimulationInitialization = self.v_simulation_initialization[i_simu_method]
            vars_list: str = ";".join(self.ui.textEdit_variables_list.toPlainText().splitlines()).strip()

            variables.extrapolate(simu_method, from_period, to_period, vars_list)

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", "Failed to extrapolate the variables.\n" + str(e))
